### ISSUE TITLE
Make ILogger implement IDisposable

### ISSUE THREAD

**Issue by ffMathy** [serilog/serilog#964]
Today, `Logger` implements `IDisposable` but `ILogger` doesn't.

This provides issues for instance with the following:

```
Logger myLoggerInstance = ...;
using(var context = myLoggerInstance.ForContext("foo", "bar")) {
    //this will not compile, since ForContext returns an ILogger and not a Logger, which is not disposable
}
```

I consider that pretty convenient!

**Comment by nblumhardt** [comment]
Thanks for the note.

I think `LogContext` is the API for your scenario:

```csharp
Logger myLoggerInstance = ...; // .Enrich.FromLogContext()
using(LogContext.PushProperty("foo", "bar")) {
    //this will not compile, since ForContext returns an ILogger and not a Logger, which is not disposable
}
```

(`ILogger`s don't have any state or resources behind them to dispose, so making them `IDisposable` would be meaningless.)

**Comment by ffMathy** [comment]
Then why is a Logger disposable?

**Comment by johnduhart** [comment]
@ffMathy `Logger` is disposable so that, as the root logger, log event sinks that implement IDisposable can be disposed. If you look at the implementation of [`Logger.ForContext`](https://github.com/serilog/serilog/blob/dev/src/Serilog/Core/Logger.cs#L96), the logger instance being created does not get passed the `_dispose` action (unlike in [LoggerConfiguration](https://github.com/serilog/serilog/blob/dev/src/Serilog/LoggerConfiguration.cs#L138-L142)), making a call to Dispose essentially a no-op.

In short, don't bother disposing `Logger` unless it's the root instance.

**Comment by nblumhardt** [comment]
It's also worth pointing out - the result returned from `ForContext()` just happens, under the covers, to be castable to `Logger` today; this isn't a documented part of the API however - it's possible a future version could return something else that may not have a `Dispose()` method on it. Hope this helps!
